File: src/project.rs
use std::collections::HashMap;

use std::{rc::Rc, cell::RefCell};
use gtk::{self, prelude::*};
use gdk::enums::key;

use gallery::Gallery;
use editor::Editor;

pub enum Item {
  Editor,
  Gallery
}

use files_tree::FilesTree;
use web_browser::WebBrowser;

pub struct View {
  item: RefCell<Item>,
  file_tree: FilesTree,
  web_browsers: RefCell<HashMap<String, WebBrowser>>,
  external_projects: RefCell<HashMap<String, Project>>
}

impl View {
  fn new() {
    let  = ;
    let tree_view = ;
    let vbox = gtk::Box::new();
    vbox.pack();
    ProjectView { overlay, project_tree,  }
  }
}

pub struct Project {
  project_views: RefCell<HashMap<String, ProjectView>>,
  notebook: gtk::NoteBook
}

impl Project {
  pub fn go_to_chapter(self, chapter_path: &str) {}
  pub fn prev_chapter(self) {}
  pub fn next_chapter(self) {}
}

File: prototype/project.lua
require "gallery"
require "editor"
require "files_tree"
require "web_browser"

local View = class()
function View:init(defaults)
  self.item = defaults.item -- Gallery or Editor
  self.files_tree = defaults.files_tree -- FilesTree
  self.floating_layer = gtk.Container()
end

Project = class()
function Project:init(defaults)
  self.views = {} -- {View}
  self.notebook = gtk.NoteBook()
  self.web_browsers = {} -- { "url" = WebBrowser }
  self.external_projects = {} -- { "url" = Project }
end

function Projecy:go_to_chapter(chapter_path)
end

function Projecy:prev_chapter()
end

function Projecy:next_chapter()
end
